library/core/src/iter/adapters/dedup.rs

```diff
@@ -0,0 +1,164 @@
+use crate::iter::{InPlaceIterable, SourceIter};
+
+/// A wrapper type around a key function.
+///
+/// This struct acts like a function which given a key function returns true
+/// if and only if both arguments evaluate to the same key.
+///
+/// This `struct` is created by [`Iterator::dedup_by_key`].
+/// See its documentation for more.
+///
+/// [`Iterator::dedup_by_key`]: Iterator::dedup_by_key
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+pub struct ByKey<F> {
+    key: F,
+}
+
+impl<F> ByKey<F> {
+    #[inline]
+    pub(crate) fn new(key: F) -> Self {
+        Self { key }
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<F, T, K> FnOnce<(&T, &T)> for ByKey<F>
+where
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    type Output = bool;
+    #[inline]
+    extern "rust-call" fn call_once(mut self, args: (&T, &T)) -> Self::Output {
+        (self.key)(args.0) == (self.key)(args.1)
+    }
+}
```

Comment:
Isn't `FnOnce` automatically inherited from `impl FnMut`?

Reply:
In practice, you're right, we only use the `FnMut` impl of this struct, but `FnMut` is defined as
```rust
pub trait FnMut<Args>: FnOnce<Args> { ... }
```
so we cannot implement `FnMut` without `FnOnce`

Reply:
I think you can delete this impl. The docs say

> `FnOnce` is implemented automatically by closures that might consume captured variables, as well as all types that implement `FnMut`,

Reply:
I tried it, it does not compile if I remove the `FnOnce` impl because `FnMut` requires it.

> `FnOnce` is implemented [for] all types that implement `FnMut`

I think this is supposed to read as "`T: FnMut` implies `T: FnOnce`", so it just means that `FnMut: FnOnce`, but not the other way around.

Reply:
I see. Shucks.